tests/src/test/java/alluxio/server/ft/MasterFailoverIntegrationTest.java

```diff
@@ -0,0 +1,127 @@
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+public final class MasterFailoverIntegrationTest extends BaseIntegrationTest {
+  private static final Logger LOG = LoggerFactory.getLogger(MasterFailoverIntegrationTest.class);
+
+  private static final String LOCAL_UFS_PATH = Files.createTempDir().getAbsolutePath();
+  private static final long DELETE_DELAY = 5 * Constants.SECOND_MS;
+
+  private MultiMasterLocalAlluxioCluster mMultiMasterLocalAlluxioCluster;
+  private FileSystem mFileSystem;
+
+  // An under file system which has slow directory deletion.
+  private static final UnderFileSystem UFS =
+      new DelegatingUnderFileSystem(UnderFileSystem.Factory.create(LOCAL_UFS_PATH)) {
+        @Override
+        public boolean deleteDirectory(String path) throws IOException {
+          CommonUtils.sleepMs(DELETE_DELAY);
+          return mUfs.deleteDirectory(path);
+        }
+
+        @Override
+        public boolean deleteDirectory(String path, DeleteOptions options) throws IOException {
+          CommonUtils.sleepMs(DELETE_DELAY);
+          return mUfs.deleteDirectory(path, options);
+        }
+      };
+
+  @ClassRule
+  public static UnderFileSystemFactoryRegistryRule sUnderfilesystemfactoryregistry =
+      new UnderFileSystemFactoryRegistryRule(new DelegatingUnderFileSystemFactory(UFS));
+
+  @Before
+  public final void before() throws Exception {
+    mMultiMasterLocalAlluxioCluster =
+        new MultiMasterLocalAlluxioCluster(2);
+    mMultiMasterLocalAlluxioCluster.initConfiguration();
+    Configuration.set(PropertyKey.USER_RPC_RETRY_MAX_DURATION, "15sec");
+    Configuration.set(PropertyKey.MASTER_THRIFT_SHUTDOWN_TIMEOUT, "30sec");
+    Configuration.set(PropertyKey.MASTER_MOUNT_TABLE_ROOT_UFS,
+        DelegatingUnderFileSystemFactory.DELEGATING_SCHEME + "://" + LOCAL_UFS_PATH);
+    mMultiMasterLocalAlluxioCluster.start();
+    mFileSystem = mMultiMasterLocalAlluxioCluster.getClient();
+  }
+
+  @After
+  public final void after() throws Exception {
+    mMultiMasterLocalAlluxioCluster.stop();
+  }
+
+  @Test
+  public void failoverJournalFencingTest() throws Exception {
+    // This test verifies that when a master fails over due to Zookeeper disconnection, outstanding
+    // threads on the master are not allowed to write to the journal.
+    AlluxioURI dir = new AlluxioURI("/dir");
+    mFileSystem.createDirectory(dir);
+    DeleteThread deleteThread = new DeleteThread(dir);
+    deleteThread.start();
+    // Give the delete thread a chance to start.
+    Thread.sleep(500);
+    mMultiMasterLocalAlluxioCluster.stopZk();
+    // Give master a chance to notice that ZK is dead and trigger failover.
+    Thread.sleep(1000);
+    mMultiMasterLocalAlluxioCluster.restartZk();
+    deleteThread.join();
+    // After failing on the original master, the delete should be retried on the new master.
+    assertFalse(mFileSystem.exists(dir));
```

Comment:
In master, this assertion would fail because the file would only be deleted on the old master.

In master, if we change this from assertFalse to mFileSystem.delete(dir), the masters will system.exit because two consecutive delete(/dir) entries get written.